Refactor the Source Type for QML Controls and more...

boutinb opened the pull request:
- Fixes some problems with ComboBox. It should also accepts source as LanguageModel.
- Refactor the SourceType so that it becomes an apart class.
- If the source of a control is the variable names, then this case should be treated as a normal kind of source. All special codes that handled the changes of dataset, can be then removed.
- Start making the QML controls independent of the MainWindow. Reference to MainWindow and DataSetPackage should be removed.
- Some code cleaning

boutinb committed:
Reactor the SourceType and more

Fixes some problems with ComboBox
Refactor the SourceType so that it becomes an apart class.
Make the column names as source type.
Start making the QML independent of the MainWindow. For this, all
changes of the dataset must be connected with the columnNames source
type.

JorisGoosen commented on Desktop/analysis/analysisform.cpp:
```diff
@@ -50,6 +50,9 @@
 using namespace std;
 
+QAbstractItemModel* AnalysisForm::columnsModel =  nullptr;
+int AnalysisForm::columnsModelRole = Qt::DisplayRole;
```
Yeah, so I really think directly querying the role like: `AnalysisForm::columnsModel->rolenames()["columnName"]` is much nicer and less chance of bugs because you do not need to remember to always make sure to update `columnsModelRole`.

I think `rolenames` might even be passed by reference, but not sure if possible.

JorisGoosen commented on Desktop/mainwindow.h:
```diff
@@ -96,6 +96,7 @@ class MainWindow : public QObject
 	bool	welcomePageVisible()	const	{ return _welcomePageVisible;	}
 	bool	checkAutomaticSync()	const	{ return _checkAutomaticSync;	}
 	QString downloadNewJASPUrl()	const	{ return _downloadNewJASPUrl;	}
+	ColumnsModel *	columnsModel()			{ return _columnsModel;			}
```
I think it would be nicer to add a `static ColumnsModel` to ColumnsModel and use that as a singleton.
I don't particularly like the whole "let's store everything in a pointer in MainWindow"-idiom we have got going on. It is easier to just directly query ColumnsModel.

Maybe this becomes less practical when we support more than 1 dataset loaded at the same time, but this doesnt seem very likely on the short term. And anyway, with `MainWindow::columnsModel()` there is the same problem.

JorisGoosen commented on Desktop/widgets/jasplistcontrol.cpp:
```diff
@@ -352,76 +118,69 @@ void JASPListControl::setUp()
+	QVector<std::pair<SourceItem*, Terms> > result; // Do not use a map in order to keep the order of the sources.
+
+	for (SourceItem* sourceItem : _sourceItems)
+		if (sourceItem->combineWithOtherModels())
```
Nicer:
```c++
for (SourceItem* sourceItem : _sourceItems)
	result.push_back(std::make_pair(sourceItem, sourceItem->combineWithOtherModels() ? _getCombinedTerms(sourceItem) : sourceItem->getTerms()));
```

JorisGoosen commented on Desktop/widgets/listmodel.cpp:
```diff
@@ -150,10 +125,11 @@ ListModel *ListModel::getSourceModelOfTerm(const Term &term)
 	ListModel* result = nullptr;
 
-	for (const std::pair<JASPListControl::SourceType *, Terms>& source : listView()->getTermsPerSource())
+	for (const auto& pair : listView()->getTermsPerSource())
```
I usually just call the pair in the iterator the names of what they are like:
```c++
for ( const auto & sourceTerms : listView()->getTermsPerSource())
	if(sourceTerms.second.contains(term) && sourceTerms.first->model())
		result = sourceTerms.first->model();
```	

It is nicer right, without the braces and the copy of the pointer?

JorisGoosen commented:
Bruno tells me this PR fixes: https://github.com/jasp-stats/INTERNAL-jasp/issues/1183

JorisGoosen commented on Desktop/widgets/listmodelfiltereddataentry.cpp:
```diff
@@ -34,12 +33,13 @@ ListModelFilteredDataEntry::ListModelFilteredDataEntry(TableViewBase * parent, Q
 }
 
-void ListModelFilteredDataEntry::dataSetChangedHandler()
+//TODO: This is not called anymore, but should be handled by termsChangedHandler
```
So is this a TODO or not?

JorisGoosen replied:
As in, it certainly looks like dataSetChangedHandler is gone now

JorisGoosen commented on Desktop/widgets/listmodelinteractionavailable.cpp:
```diff
@@ -20,33 +20,31 @@
 	{
-		ListModel* sourceModel = source.first->model;
-		const Terms& terms = source.second;
+		SourceItem* sourceItem = pair.first;
```
Also here it would make the whole code that much nicer to look at.

JorisGoosen replied:
Just give the paris descriptive names instead of copying shit :p

JorisGoosen commented on Desktop/widgets/listmodellabelvalueterms.cpp:
```diff
@@ -132,23 +116,25 @@ void ListModelLabelValueTerms::_setLabelValues(const JASPListControl::LabelValue
+		labelValuePairs.push_back(std::make_pair(_listView->placeholderText(), ""));
 
-	for (const std::pair<JASPListControl::SourceType *, Terms>& source : _listView->getTermsPerSource())
+	for (const auto& pair : listView()->getTermsPerSource())
```
Or maybe do it like `void		applyToAll(std::function<void(Analysis *analysis)> applyThis);`
The you could just have it like:
` void forAllSourceTerms(std::function<void(SourceItem * item, const Terms & terms)> callThis);`

That way all of these `for ... copy first and second, do stuff` loops can be so much nicer than what Youve been adding.

JorisGoosen replied:
You could even make one where you check if it has a model and make a `forAllSourceTermsWithModel(std::function<void(SourceItem * item, ListModel * model)> callThis)` or something like that

boutinb replied:
Nice. I'll do this

JorisGoosen commented on Desktop/widgets/sourceitem.cpp:
```diff
@@ -0,0 +1,406 @@
+	}
+	else if (_nativeModel)
+	{
+		connect(_nativeModel, SIGNAL(dataChanged(const QModelIndex &, const QModelIndex &, const QVector<int> &)),	this,	SLOT(nativeModelChanged()) );
```
New style `connect`'s weren't possible I suppose?

JorisGoosen replied:
(You can refer to functions like `&SuperClass::Function`)

boutinb replied:
Because of the parameters, I chose this case. But I could use a lambda instead, and adding the sourceItem as context to be sure that the connection is destroyed when the SourceItem is destroyed.

JorisGoosen replied:
But the normal connect also drops the parameters right?
As in, I don't see why this version would be better?

JorisGoosen replied:
By that I mean, why not do:
`connect(_nativeModel, &QAbstractItemModel::dataChanged, this, &SourceItem::nativeModelChanged);` should work I think.

boutinb replied:
That's true. At some point I thought that would not work, but it works indeed!

JorisGoosen replied:
Yeah, connects automatically drop arguments/parameters from the signal if they are not needed/accepted by the slot.

JorisGoosen reviewed (changes requested):
Phew, you have been working hard!

I think a few changes could make things a bit better. But I see nothing that would break necessarily, just things that would make the code easier to maintain.

I'll be building it now and testing.

JorisGoosen commented:
Ok, testing showed my that replacing a variablename in the data through spreadsheeteditor only removes it from variableslists... It does not replace them...

EDIT: that means that if `V1` is used in Descriptives and I rename it to `somethingElse` it gets removed from the list. Not what we want.

boutinb commented:
The problem is that I have now only the ColumnModel, and it does not give such a signal that tells that the column name is changed.
This should be solved with the DataSetPackage solution we discussed.

boutinb committed:
Add Joris' requests.

JorisGoosen commented:
> 
> 
> The problem is that I have now only the ColumnModel, and it does not give such a signal that tells that the column name is changed.
> This should be solved with the DataSetPackage solution we discussed.

Doesn't it send a `dataChanged` when the columnName changes?
I think so.

JorisGoosen commented on Desktop/analysis/analysisform.cpp:
```diff
@@ -782,6 +782,22 @@ QString AnalysisForm::metaHelpMD() const
+		while (i.hasNext())
+		{
+			i.next();
+			if (i.value() == "columnName")
```
If you've found it you should `break;` the `while` loop.

JorisGoosen replied:
And... QHash is a lookup table, so you can just do `model->roleNames()["columnName"]` which I think I mentioned before.

Doing that also means you can remove `AnalysisForm::_columnsModelRole` entirely and look it up every time.

JorisGoosen replied:
Ps. see if rolenames can return a `QHash &` because otherwise it is copying the whole thing all the time. Not *so bad* but also not so good.

boutinb replied:
RoleNames is a QHash<int, QByteArrea>, so roleNames()["columnName"] won't work.
I'll add  a break in the loop, but as we can't have the role directly with a lookup, I'll keep the _columnsModelRole. This name don't change.

And columnsModel does not send dataChange signal. I will handle this issue with DataSetPackage rewrite.

JorisGoosen replied:
Aah, well `columnsModel` should send that, but I guess I was lazy when implementing it :p

And ah, yeah, ok that is the wrong way round with the QHash... Ok, then the copy is reasonable.

JorisGoosen reviewed (changes requested):
Looks much better already, but I don't get why you aren't using QHash properly.

boutinb committed:
Last small changes

JorisGoosen commented:
I'll go test if variablenames are now properly replaced.

JorisGoosen commented on Desktop/widgets/sourceitem.cpp:
```diff
@@ -102,10 +96,10 @@ void SourceItem::_setUp()
-		connect(_nativeModel, &QAbstractItemModel::modelReset,			this,	&SourceItem::nativeModelChanged);
+		connect(_nativeModel, &QAbstractItemModel::dataChanged,			this,	[this]() { _listControl->model()->sourceTermsChanged(); });
+		connect(_nativeModel, &QAbstractItemModel::rowsInserted,		this,	[this]() { _listControl->model()->sourceTermsChanged(); });
+		connect(_nativeModel, &QAbstractItemModel::rowsRemoved,			this,	[this]() { _listControl->model()->sourceTermsChanged(); });
```
Ok, this is not necessarily a complaint, or request for change but I am not sure why you did:
`connect(_nativeModel, &QAbstractItemModel::dataChanged,			this,	[this]() { _listControl->model()->sourceTermsChanged(); });`
When you could also do:
`connect(_nativeModel, &QAbstractItemModel::dataChanged,			this,	_listControl->model(), &ListModel::sourceTermsChanged());` 
(That is, if `_listControl->model()` is a ListModel.

JorisGoosen commented:
Ok, so replacing variablenames still breaks...

JorisGoosen commented:
Can this be reviewed again?

boutinb committed:
More fixes & rewrite

Change column variable does not remove anymore the variable from an
assigned list.
Remove the termsReset signal and uses the Qt model signals
Centralize the use and update of the terms in ListModel at one place.

boutinb commented:
Now you can review it. I've changed more things...
